### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by robin-leach**
I wasn't able to finish this exercise. I got up to starting the selenium setup, but despite my best efforts I couldn't get the environment variables to load properly in the test app, and so couldn't get the end to end tests to work.

### CODE REVIEW

File: e2e_tests/end_to_end_test.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    # construct the new application
+    application = app.create_app()
+
+    log.debug(os.getenv('TRELLO_KEY'))
+
+    # start the app in its own thread.
+    thread = Thread(target=lambda: application.run(use_reloader=False))
+    thread.daemon = True
+    thread.start()
+    yield app
```
Comment (Jonesey13) [inline e2e_tests/end_to_end_test.py:31]: Undefined variable (do you mean `application`?)

File: helpers/index_test.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
+import pytest
```
Comment (Jonesey13) [inline helpers/index_test.py:1]: Not added to `requirements.txt`

File: helpers/index_test.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+from datetime import datetime, timedelta
+from freezegun import freeze_time
```
Comment (Jonesey13) [inline helpers/index_test.py:3]: Not added to `requirements.txt`

File: test_app.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from app import create_app
+from entity.trello_list import TrelloList
+from entity.trello_card import TrelloCard
+
+import consts
+
+
+@pytest.fixture
+def client():
+    file_path = find_dotenv('test_env_file')
```
Comment (Jonesey13) [inline test_app.py:14]: You'll want `find_dotenv('test_env_file.txt')` although this will clash with the e2e test run

Comment (robin-leach) [reply test_app.py:14]: Why will it clash?

File: .gitignore [hunk 5]
```diff
@@ -114,3 +114,7 @@ env/
 
 ### Configuration files ###
 .env
+
+### Geckodriver
+geckodriver.exe
```
Comment (hugh-emerson) [inline .gitignore:119]: Good to gitignore this but adding things to gitignore won't actually remove them from the repo if they were already being tracked. You need to delete the geckodriver.exe file from the repo yourself.

File: README.md [hunk 6]
```diff
@@ -23,3 +23,10 @@ You should see output similar to the following:
  * Debugger PIN: 226-556-590
 ```
 Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser to view the app.
+
+
+## Testing
+
+To run all tests, run `python -m pytest`
+To run integration tests, run `python -m pytest test_app.py`
+To run end-to-end tests, run `python -m pytest e2e_tests/end_to_end_test.py`
```
Comment (hugh-emerson) [inline README.md:32]: You should also mention the prerequisites - geckodriver.exe (once you delete it from the repo) and Firefox

File: README.md [hunk 7]
```diff
@@ -23,3 +23,10 @@ You should see output similar to the following:
  * Debugger PIN: 226-556-590
 ```
 Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser to view the app.
+
+
+## Testing
+
+To run all tests, run `python -m pytest`
```
Comment (hugh-emerson) [inline README.md:30]: Is there any reason to run `python -m pytest` rather than just `pytest`?

**Review by hugh-emerson — CHANGES_REQUESTED** [review]
See my comments for an explanation of what's going wrong. Let me know via Slack or book a 1-to-1 if you still can't get the tests working.

**Comment by hugh-emerson** [comment]
Markups look good to me :+1: Merge away!

**Pull request merged**